cpp/velox/compute/VeloxPlanConverter.cc

```diff
@@ -251,59 +252,17 @@ class VeloxPlanConverter::WholeStageResIter {
   void toRealArrowBatch(const RowVectorPtr& rv, uint64_t numRows,
                         const RowTypePtr& outTypes,
                         std::shared_ptr<arrow::RecordBatch>* out) {
-    uint32_t colNum = outTypes->size();
-    std::vector<std::shared_ptr<arrow::Array>> outArrays;
-    outArrays.reserve(colNum);
-    std::vector<std::shared_ptr<arrow::Field>> retTypes;
-    retTypes.reserve(colNum);
-    for (uint32_t idx = 0; idx < colNum; idx++) {
-      arrow::ArrayData outData;
-      outData.length = numRows;
-      auto vec = rv->childAt(idx);
-      // FIXME: need to release this.
-      ArrowArray arrowArray;
-      exportToArrow(vec, arrowArray, veloxPool_.get());
-      outData.buffers.resize(arrowArray.n_buffers);
-      outData.null_count = arrowArray.null_count;
-      // Validity buffer
-      std::shared_ptr<arrow::Buffer> valBuffer = nullptr;
-      if (arrowArray.null_count > 0) {
-        arrowArray.buffers[0];
-        // FIXME: set BitMap
-      }
-      outData.buffers[0] = valBuffer;
-      auto colType = outTypes->childAt(idx);
-      auto colArrowType = toArrowType(colType);
-      // TODO: use the names in RelRoot.
-      auto colName = "res_" + std::to_string(idx);
-      retTypes.emplace_back(arrow::field(colName, colArrowType));
-      outData.type = colArrowType;
-      if (colType->isPrimitiveType() && !facebook::velox::substrait::isString(colType)) {
-        auto dataBuffer = std::make_shared<arrow::Buffer>(
-            static_cast<const uint8_t*>(arrowArray.buffers[1]),
-            facebook::velox::substrait::bytesOfType(colType) * numRows);
-        outData.buffers[1] = dataBuffer;
-      } else if (facebook::velox::substrait::isString(colType)) {
-        auto offsets = static_cast<const int32_t*>(arrowArray.buffers[1]);
-        int32_t stringDataSize = offsets[numRows];
-        auto valueBuffer = std::make_shared<arrow::Buffer>(
-            static_cast<const uint8_t*>(arrowArray.buffers[2]), stringDataSize);
-        auto offsetBuffer = std::make_shared<arrow::Buffer>(
-            static_cast<const uint8_t*>(arrowArray.buffers[1]),
-            sizeof(int32_t) * (numRows + 1));
-        outData.buffers[1] = offsetBuffer;
-        outData.buffers[2] = valueBuffer;
-      }
-      std::shared_ptr<arrow::Array> outArray =
-          MakeArray(std::make_shared<arrow::ArrayData>(std::move(outData)));
-      outArrays.emplace_back(outArray);
-      // int ref_count = vec->mutableValues(0)->refCount();
+    ArrowArray cArray{};
+    ArrowSchema cSchema{};
+    exportToArrow(rv, cArray, veloxPool_.get());
+    exportToArrow(outTypes, cSchema);
+    arrow::Result<std::shared_ptr<arrow::RecordBatch>> batch =
+        arrow::ImportRecordBatch(&cArray, &cSchema);
```

Comment:
Verified locally, and found no performance regression on TPC-H q1/q6 on sf1000.
Just a question: will this implementation use memcopy to create a batch, or just change the pointer in the data buffer?

Reply:
It appears to copy string data in [velox code](https://github.com/oap-project/velox/blob/main/velox/vector/arrow/c/Bridge.cpp#L210).

The memcopy should more or less bring overhead although in usual case it can be too fast to conduct real regressions.

The previous codes is more like a workaround to me however we can still track this part's performance in future in case we need another solution to achieve the same goal.

Reply:
Thanks for the answering. Actually I'm asking the process from abi array to record batch. Would memcopy happen here?
```
arrow::ImportRecordBatch(&cArray, &cSchema)
```

Reply:
Ideally this function doesn't copy data. It's just a regular Arrow ABI struct importing.